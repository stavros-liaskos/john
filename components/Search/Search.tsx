import { SearchProps } from './Search.types';

const Search = ({ i18n }: SearchProps) => {
  if (!i18n || !i18n.button || !i18n.label) {
    return null;
  }

  const handleSubmit = () => {};

  return (
    <form className="border border-indigo-600" noValidate onSubmit={handleSubmit}>
      <label htmlFor="search">{i18n.label}</label>
      <input type="text" name="search" />
      <button data-testid="search-button">{i18n.button}</button>
    </form>
  );
};

export default Search;
