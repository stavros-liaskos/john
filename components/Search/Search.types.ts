export interface SearchI18n {
  button: string;
  label: string;
}

export interface SearchProps {
  i18n: SearchI18n;
  handleSearch: Function;
}
