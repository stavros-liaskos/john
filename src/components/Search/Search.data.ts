import { SearchI18n } from './Search.types';

export const searchI18n: SearchI18n = {
  label: 'Here goes the artist...',
  button: 'Search',
  searchList: {
    btnTxt: 'Follow',
    noArtists: 'No artists found',
  },
};
