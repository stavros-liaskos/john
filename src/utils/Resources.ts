import { components } from '../types/schema';

interface IResources {
  cacheMap: Record<string, unknown>;
  fetch: (url: string) => void;
}

class Resources implements IResources {
  private cacheMap;

  constructor() {
    this.cacheMap = new Map();
  }

  fetch(url) {
    if (this.cacheMap.has(url)) {
      return this.cacheMap.get(url);
    }

    const promise = fetch(url, {
      method: 'GET',
      credentials: 'include',
    })
      .then(response => {
        return response.json();
      })
      .then((data: components['schemas']['FollowedArtistsResponse']) => {
        this.cacheMap.set(url, data);
        return data;
      });

    throw promise;
  }
}

const resources = new Resources();

export default resources;
