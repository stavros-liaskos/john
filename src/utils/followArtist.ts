import { components } from '../types/schema';
import Endpoints from '../types/endpoints';

async function followArtist(artist: components['schemas']['SearchResultArtistDto'], cb: () => void) {
  const headers = new Headers({ 'Content-Type': 'application/json' });
  await fetch(Endpoints.Follow, {
    method: 'POST',
    credentials: 'include',
    headers,
    body: JSON.stringify(artist),
  })
    .then(() => {
      console.log(`${artist.name} followed successfully`);
    })
    .catch(error => {
      console.error('Error:', JSON.stringify(error));
    })
    .finally(() => {
      cb();
    });
}
export default followArtist;
