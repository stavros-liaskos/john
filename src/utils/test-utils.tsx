import React from 'react';
import { render, act } from '@testing-library/react';
import ThemeProvider from '../contexts/Theme/ThemeProvider';
import { UserProvider } from '@auth0/nextjs-auth0/client';
import ArtistsListProvider from '../contexts/ArtistsList/ArtistsListProvider';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <UserProvider>
        <ArtistsListProvider>{children}</ArtistsListProvider>
      </UserProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: { [key: string]: unknown }) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const renderWithAct = async (ui: React.ReactElement, options?: { [key: string]: unknown }) => {
  const component = render(ui, { wrapper: AllTheProviders, ...options });

  const promise = Promise.resolve();
  await act(() => promise);

  return component;
};

export const beforeEachTest = () => {
  jest.resetModules();
  jest.clearAllMocks();
  jest.restoreAllMocks();
  localStorage.clear();
};
