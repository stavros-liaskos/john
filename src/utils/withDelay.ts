export default function withDelay(cb: () => void, delay = 1000) {
  return setTimeout(() => {
    cb();
  }, delay);
}
