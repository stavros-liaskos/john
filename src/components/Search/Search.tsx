import { SearchProps } from './Search.types';
import React, { useState } from 'react';
import Button from '../Button/Button';
import { components } from '../../types/schema';
import FormInput from '../FormInput/FormInput';
import ArtistsList from '../ArtistsList/ArtistsList';
import followArtist from '../../utils/followArtist';
import { useArtistsListContext } from '../../contexts/ArtistsList/ArtistsListContext';
import Close from '../Icons/close';
import HandGlass from '../Icons/handGlass';

const Search: React.FunctionComponent<SearchProps> = ({ i18n }) => {
  const { getFollowedArtists } = useArtistsListContext();
  const [artistLoading, setArtistLoading] = useState<number>(0);
  const [results, setResults] = useState<components['schemas']['SearchResultArtistDto'][] | null>(null);
  const [disabled, setDisabled] = useState<boolean>(false);

  if (!i18n?.label) {
    return null;
  }
  return (
    <div className="relative flex lg:justify-center items-center flex-none h-16 md:h-20 md:border-b-2 rr-border w-full">
      <FormInput handleAction={handleSearch} i18n={i18n} actionEventTrigger={'onSubmit'}>
        {results && (
          <button onClick={() => setResults(null)}>
            <Close />
          </button>
        )}
        <Button className="btn-large md:ml-52" type="submit" disabled={disabled} loading={disabled} aria-label="search">
          <HandGlass />
        </Button>
      </FormInput>

      {results && (
        <div className="absolute px-3 bg-slate-100 dark:bg-gh-darkly border-2 rr-border top-14 md:top-16 w-full z-10">
          <ArtistsList
            i18n={i18n.searchList}
            artistsList={results}
            onButtonClick={artist => handleFollow(artist)}
            artistLoading={artistLoading}
          />
        </div>
      )}
    </div>
  );

  async function handleSearch(inputValue: string) {
    inputValue &&
      (await fetch(`/artist/search?${new URLSearchParams({ pattern: inputValue })}`, {
        method: 'GET',
        credentials: 'include',
      })
        .then(res => res.json())
        .then(result => {
          getFollowedArtists();
          return setResults(result.artists);
        })
        .catch(console.error));
  }

  async function handleFollow(artist: components['schemas']['SearchResultArtistDto']) {
    artist?.id && setArtistLoading(artist.id);
    setDisabled(true);

    const finallyCb = () => {
      artistLoading && setArtistLoading(0);
      setResults(null);
      setDisabled(false);
    };
    await followArtist(artist, finallyCb);
  }
};
Search.whyDidYouRender = true;
export default Search;
