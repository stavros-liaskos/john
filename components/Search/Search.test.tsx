import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import Search from './Search';
import { searchI18n } from './Search.data';
import { beforeEachTest, renderWithAct } from '../../utils/test-utils';
import fetchMock from 'jest-fetch-mock';
import followedArtists from '../../mocks/responses/followed-artists.json';
import artistSearch from '../../mocks/responses/artist-search.json';

const setup = async () => {
  const { container, findByRole } = await renderWithAct(<Search i18n={searchI18n} />);
  const searchBtn = await findByRole('button');
  const input = await findByRole('textbox');

  return {
    container,
    searchBtn,
    input,
  };
};

describe('Search', () => {
  beforeEach(() => {
    beforeEachTest();

    fetchMock.mockIf(/^https?:\/\/release-raccoon.com.*$/, req => {
      if (req.url.endsWith('/me/followed-artists')) {
        return Promise.resolve({ body: JSON.stringify({ json: followedArtists }) });
      } else {
        return Promise.reject({ status: 404 });
      }
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders without data without crashing', () => {
    // @ts-ignore
    render(<Search />);
  });

  it('should have the query that the user types in the input', async () => {
    const { input } = await setup();

    fireEvent.change(input, { target: { value: 'Nels Cline' } });
    expect(input).toHaveValue('Nels Cline');
  });

  it('should handle the search action of the user', async () => {
    const { container, input, searchBtn } = await setup();

    fetchMock.mockIf(/^https?:\/\/release-raccoon.com.*$/, req => {
      if (req.url.endsWith('Nels+Cline')) {
        return Promise.resolve({ body: JSON.stringify({ json: artistSearch }) });
      } else {
        return Promise.reject({ status: 404 });
      }
    });

    fireEvent.change(input, { target: { value: 'Nels Cline' } });
    await act(() => {
      fireEvent.click(searchBtn);
    });
    // TODO add nock
    // expect(handleSearch).toHaveBeenCalledWith('Nels Cline');
    expect(container).toMatchSnapshot();
  });
});
