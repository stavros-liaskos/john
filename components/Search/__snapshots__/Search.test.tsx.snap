// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should handle the search action of the user 1`] = `
<div>
  <div
    class="relative flex lg:justify-center items-center mb-2 h-20 md:h-40 border-b-2 rr-border w-full"
  >
    <form
      class="flex justify-between md:justify-between items-stretch h-10 w-full"
      novalidate=""
    >
      <input
        class="mr-4 px-2 min-m-lg border-b-2 rr-border dark:bg-gh-darkly rr-text w-full md:w-2/3"
        name="search"
        placeholder="Here goes the artist..."
        type="text"
      />
      <button
        class="btn flex btn-large"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</div>
`;
